lib/devices_test.go

```diff
@@ -0,0 +1,197 @@
+	mockRefresh([]Device{d1})
+	changed, merged := deviceDiff([]Device{d1})
+	if len(merged) != 1 {
+		t.Errorf("deviceDiff no change failed: expected 1 device, got %d", len(merged))
+	}
+	if changed {
+		t.Errorf("deviceDiff no change failed: expected changed=false, got true")
+	}
+}
+
+func TestDeviceDiff_AddAndRemove(t *testing.T) {
+	d1 := Device{Path: []int{1}, Name: "A", VendorId: "v1", ProductId: "p1", Speed: "High", Bus: 1, State: StateNormal}
+	d2 := Device{Path: []int{2}, Name: "B", VendorId: "v2", ProductId: "p2", Speed: "High", Bus: 1, State: StateNormal}
+	d3 := Device{Path: []int{3}, Name: "C", VendorId: "v3", ProductId: "p3", Speed: "High", Bus: 1, State: StateNormal}
+	mockRefresh([]Device{d1, d2})
+	changed, merged := deviceDiff([]Device{d2, d3})
+	if !changed {
+		t.Errorf("deviceDiff add+remove failed: expected changed=true, got false")
+	}
+	if !hasState(merged, StateAdded) {
+		t.Errorf("deviceDiff add+remove failed: expected StateAdded in merged devices")
+	}
+	if !hasState(merged, StateRemoved) {
+		t.Errorf("deviceDiff add+remove failed: expected StateRemoved in merged devices")
+	}
+}
+
+func hasState(devs []Device, state LogState) bool {
+	for _, d := range devs {
+		if d.State == state {
+			return true
+		}
+	}
+	return false
+}
+
+func TestIsChild(t *testing.T) {
+	parent := TreeNode{
+		Device: Device{
+			Path: []int{1, 2},
+			Bus:  1,
+		},
+	}
+	child := TreeNode{
+		Device: Device{
+			Path: []int{1, 2, 3},
+			Bus:  1,
+		},
+	}
+	notChild := TreeNode{
+		Device: Device{
+			Path: []int{1, 3, 4},
+			Bus:  1,
+		},
+	}
+	// Should be true: child path extends parent path by one element and matches up to parent length
+	if !isChild(parent, child) {
+		t.Errorf("isChild failed: expected true for child %+v of parent %+v", child, parent)
+	}
+	// Should be false: notChild path does not match parent path
+	if isChild(parent, notChild) {
+		t.Errorf("isChild failed: expected false for notChild %+v of parent %+v", notChild, parent)
+	}
+	// Should be false: same path length
+	if isChild(parent, parent) {
+		t.Errorf("isChild failed: expected false for same node")
+	}
+	// Should be false: child path is shorter
+	if isChild(child, parent) {
+		t.Errorf("isChild failed: expected false when parent is actually child")
+	}
+}
+
+func TestBuildDeviceTreeAndIsChild(t *testing.T) {
+	d1 := Device{Path: []int{}, Name: "Root", VendorId: "v1", ProductId: "p1", Speed: "High", Bus: 1}
+	d2 := Device{Path: []int{1}, Name: "Child", VendorId: "v2", ProductId: "p2", Speed: "High", Bus: 1}
+	d3 := Device{Path: []int{1, 2}, Name: "Grandchild", VendorId: "v3", ProductId: "p3", Speed: "High", Bus: 1}
+	tree := BuildDeviceTree([]Device{d1, d2, d3})
+	if len(tree) != 1 {
+		t.Errorf("BuildDeviceTree failed: expected 1 root, got %d, tree: %+v", len(tree), tree)
+	}
+	if len(tree) > 0 && len(tree[0].Children) != 1 {
+		t.Errorf("BuildDeviceTree failed: expected 1 child for root, got %d, tree: %+v", len(tree[0].Children), tree)
+	}
+}
```

Comment:
It would be more comforting to know you've tested the entire tree structure matches expected state. You could e.g. compare vs `fmt.Sprintf("%#v")`, or build a string of each tree item + their indentation level and compare against an expected value.